Sources/MapboxCoreNavigation/NavigationRouter.swift

```diff
@@ -0,0 +1,179 @@
+import Foundation
+import MapboxDirections
+import MapboxNavigationNative
+
+public class NavigationRouter {
+    
+    public typealias RequestId = UInt64
+    
+    public struct RoutingRequest {
+        public let id: RequestId
+        
+        // Intended retain cycle to prevent deallocating. `RoutingRequest` will be deleted once request completes.
+        let router: NavigationRouter
+        
+        public func cancel() {
+            router.finish(request: id)
+        }
+    }
+    
+    public enum RouterSource {
+        case online
+        case offline
+        case hybrid
+        
+        var nativeSource: RouterType {
+            switch self {
+            case .online:
+                return .online
+            case .offline:
+                return .onboard
+            case .hybrid:
+                return .hybrid
+            }
+        }
+    }
+    
+    public private(set) var activeRequests: [RequestId : RoutingRequest] = [:]
+    public private(set) var source: RouterSource
+    private var router: RouterInterface
+    private var settings: NavigationSettings
+    
+    public init(_ source: RouterSource = .hybrid, settings: NavigationSettings = .shared) {
+        self.source = source
+        self.settings = settings
+        
+        let factory = NativeHandlersFactory(tileStorePath: settings.tileStoreConfiguration.navigatorLocation.tileStoreURL?.path ?? "",
+                                            credentials: settings.directions.credentials)
+        self.router = MapboxNavigationNative.RouterFactory.build(for: source.nativeSource,
+                                                                 cache: factory.cacheHandle,
+                                                                 historyRecorder: factory.historyRecorder)
+    }
+    
+    deinit {
+        print(">>> dealloc")
+    }
+    
+    fileprivate func finish(request id: RequestId) {
+        router.cancelRequest(forToken: id)
+        activeRequests[id] = nil
+        print(">>> finished id: \(id)")
+    }
+    
+    public func doRequest<ResponseType: Codable>(options: DirectionsOptions,
+                                                 completion: @escaping (Result<ResponseType, DirectionsError>) -> Void) -> RequestId {
```

Comment:
Controlling the `completionQueue` that defaults to .main, would be a good customization point.

Reply:
`NavigationRouter` is an alternative to our regular `Directions` response wrappers which always call their completions on main thread. I assume that will be also expected true for `NavigationNavigator`, and it will also be mentioned in the docs.